codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithygo/awssdk/shapevisitor/DafnyToAwsSdkShapeVisitor.java

```diff
@@ -556,16 +549,86 @@ public String doubleShape(DoubleShape shape) {
 
   @Override
   public String unionShape(UnionShape shape) {
-    writer.addImportFromModule("github.com/dafny-lang/DafnyRuntimeGo", "dafny");
-    return """
-    func () types.%s {
-    _ = val
-    return nil
-    }()""".formatted(context.symbolProvider().toSymbol(shape).getName());
+    writer.addImportFromModule("github.com/dafny-lang/DafnyStandardLibGo", "Wrappers");
+    var nilCheck = "";
+    if (this.isOptional) {
+      final String unAssertDataSource = dataSource.startsWith("input.(") ? "input" : dataSource;
+      if (this.isPointable) {
+          nilCheck = "if %s == nil { return nil }".formatted(unAssertDataSource);
+      } else {
+          nilCheck = "if %s == nil { return union}".formatted(unAssertDataSource);
+      }
+    }
+    final String functionInit = """
+        func() %s {
+            var union %s
+            %s
+        """.formatted(
+            SmithyNameResolver.getSmithyTypeAws(serviceTrait, context.symbolProvider().toSymbol(shape), true),
+            SmithyNameResolver.getSmithyTypeAws(serviceTrait, context.symbolProvider().toSymbol(shape), true),
+            nilCheck
+    );
+    final StringBuilder eachMemberInUnion = new StringBuilder();
+    for (final var member : shape.getAllMembers().values()) {
+        final Shape targetShape = context.model().expectShape(member.getTarget());
+        final String memberName = context.symbolProvider().toMemberName(member);
+        // unwrap union type, assert it then convert it to its member type with Dtor_ (example: Dtor_BlobValue()). unionDataSource is not a wrapper object until now.
+        String unionDataSource = dataSource + ".Dtor_" + memberName.replace(shape.getId().getName() + "Member", "") + "()";
```

Comment:
`.concat()` ?

Reply:
Done here (including a lot of other stuffs) https://github.com/smithy-lang/smithy-dafny/commit/75c8e6a384e4c085e161578cfb682151c7d02e8e